tmail-backend/jmap/extensions/src/main/scala/com/linagora/tmail/james/jmap/method/PublicAssetSetMethod.scala

```diff
@@ -92,7 +97,10 @@ class PublicAssetSetCreatePerformer @Inject()(val publicAssetRepository: PublicA
   private def createPublicAssets(mailboxSession: MailboxSession,
                                  creationRequest: PublicAssetSetCreationRequest): SMono[PublicAssetStorage] =
     generatePublicAssetCreationRequest(creationRequest, mailboxSession)
-      .flatMap(publicAsset => SMono(publicAssetRepository.create(mailboxSession.getUser, publicAsset)))
+      .flatMap(publicAsset => SMono(publicAssetRepository.create(mailboxSession.getUser, publicAsset))
+        .flatMap(publicAssetStorage =>
+          SMono(currentQuotaManager.increase(new QuotaOperation(userQuotaRootResolver.forUser(mailboxSession.getUser), QuotaCountUsage.count(1), QuotaSizeUsage.size(publicAssetStorage.size))))
+            .`then`(SMono.just(publicAssetStorage))))
```

Comment:
:-(

Several thing wrong here.

-> **Layers are leaking!** I would expect the quota topic to be managed by the storage layer at the repository level. It should not actually be called in the presentation layer. The presentation layer shall just be *presenting* the error from the storage layer.

Why?

 - This makes the presentation layer more complex.
 - If the storage layer is reused for other purposes, then the checks defined here would not be applied to other protocols.
 
  -> `hammer screw symptom`. It's not because we just bought a hammer that we need to use it all the time.

Here I expect the cardinality of PublicAsset to be small thus at creation we could list all users public assets and then decide if we can add one more or not. No need to use the fancy currentQuotaManager for this.

 -> If we actually create a projection (what I would like to avoid cf above) then we need to hand a tool for correcting it / recomputing it. So if the team disregard the earlier remark, at the very least it shall be plugged onto the currentQuotaRecomputing stuff, and tested. At the very least in another ticket included in the sprint.

Reply:
implemented "list all users public assets and then decide if we can add one more or not"